File: android/app/src/main/java/javax/security/sasl/SaslException.java
package javax.security.sasl;

import java.io.IOException;

/**
 * Minimal Android stub of {@code javax.security.sasl.SaslException}.
 * Required so MongoDB driver SCRAM classes can load and throw SASL failures.
 */
public class SaslException extends IOException {
    private static final long serialVersionUID = 4579784287902078538L;

    private Throwable exception;

    public SaslException() {
        super();
    }

    public SaslException(String detail) {
        super(detail);
    }

    public SaslException(String detail, Throwable ex) {
        super(detail);
        if (ex != null) {
            initCause(ex);
        }
        this.exception = ex;
    }

    @Override
    public Throwable getCause() {
        return exception;
    }

    @Override
    public Throwable initCause(Throwable cause) {
        super.initCause(cause);
        this.exception = cause;
        return this;
    }
}

File: android/app/src/main/java/javax/security/sasl/SaslException.kt
package javax.security.sasl

import java.io.IOException

/**
 * Minimal Android stub of `javax.security.sasl.SaslException`.
 * Required so MongoDB driver SCRAM classes can load and throw SASL failures.
 */
class SaslException : IOException {
    private var exception: Throwable? = null

    constructor() : super()

    constructor(detail: String?) : super(detail)

    constructor(detail: String?, ex: Throwable?) : super(detail) {
        if (ex != null) {
            initCause(ex)
        }
        this.exception = ex
    }

    override fun getCause(): Throwable? = exception

    override fun initCause(cause: Throwable?): Throwable {
        super.initCause(cause)
        this.exception = cause
        return this
    }

    companion object {
        private const val serialVersionUID = 4579784287902078538L
    }
}
